### PR TITLE
Added Node class which will be used in the A* algorithm

### PR DESCRIPTION
**Pull request by andrewoid**
A video describing the algorithm can be viewed here https://www.youtube.com/watch?v=-L-WgKMFuhE

Video of A* in action be be viewed here.
https://github.com/andrewoid/SmartSnake/assets/121566/d84d3e2c-c717-466e-949f-c590020ac9ba

### CODE REVIEW

**Review by chanacham — COMMENTED** [review]
The up arrow sends the snake down and the down arrow sends the snake up. Also, when the game quits the screen goes blank- there should be a way of handling that and restarting the game

**Comment by MC-Student** [comment]
The reason up and down are reversed is because the coordinates start from the top left - x behaves normally but when y increases,  you actually move towards the top of the screen which has the y coordinate of 0.

**Comment by MC-Student** [comment]
h = height of gui
Points should be set to (x, h - y)

**Review by MC-Student — COMMENTED** [review]
h = height of gui
Points should be set to (x, h - y) to solve vertical direction problem

File: src/main/java/smartsnake/astar/Node.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        return "Node{" +
+                "location=" + location +
+                ", fromStart=" + fromStart +
+                ", fromEnd=" + fromEnd +
+                '}';
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
```
Comment (lkenigsb) [inline src/main/java/smartsnake/astar/Node.java]: Do not use == when comparing objects

Comment (andrewoid) [reply src/main/java/smartsnake/astar/Node.java]: The code for equals was generated by IntelliJ and is correct because it does further equality checks.

**Review by ElkySchles — APPROVED** [review]

**Review by MC-Student — APPROVED** [review]

**Review by elkyhalberstam — APPROVED** [review]

**Pull request merged**